hugr-core/src/export.rs

```diff
@@ -0,0 +1,726 @@
+//! Exporting HUGR graphs to their `hugr-model` representation.
+use crate::{
+    extension::ExtensionSet,
+    ops::OpType,
+    types::{
+        type_param::{TypeArgVariable, TypeParam},
+        type_row::TypeRowBase,
+        CustomType, EdgeKind, FuncTypeBase, MaybeRV, PolyFuncTypeBase, RowVariable, SumType,
+        TypeArg, TypeBase, TypeEnum,
+    },
+    Direction, Hugr, HugrView, IncomingPort, Node, Port, PortIndex,
+};
+use bumpalo::{collections::Vec as BumpVec, Bump};
+use hugr_model::v0::{self as model};
+use indexmap::IndexSet;
+use smol_str::ToSmolStr;
+
+pub(crate) const OP_FUNC_CALL_INDIRECT: &str = "func.call-indirect";
+const TERM_PARAM_TUPLE: &str = "param.tuple";
+
+/// Export a [`Hugr`] graph to its representation in the model.
+pub fn export_hugr<'a>(hugr: &'a Hugr, bump: &'a Bump) -> model::Module<'a> {
+    let mut ctx = Context::new(hugr, bump);
+    ctx.export_root();
+    ctx.module
+}
+
+/// State for converting a HUGR graph to its representation in the model.
+struct Context<'a> {
+    /// The HUGR graph to convert.
+    hugr: &'a Hugr,
+    /// The module that is being built.
+    module: model::Module<'a>,
+    /// Mapping from ports to link indices.
+    /// This only includes the minimum port among groups of linked ports.
+    links: IndexSet<(Node, Port)>,
+    bump: &'a Bump,
+}
+
+impl<'a> Context<'a> {
+    pub fn new(hugr: &'a Hugr, bump: &'a Bump) -> Self {
+        let mut module = model::Module::default();
+        module.nodes.reserve(hugr.node_count());
+
+        Self {
+            hugr,
+            module,
+            bump,
+            links: IndexSet::new(),
+        }
+    }
+
+    pub fn export_root(&mut self) {
```

Comment:
Why is the logic for the root module of the Hugr different from the node's?

This seems to assume the hugr root is a Module.
As far as I understand, passing a DFG-rotted hugr here will fail when calling `self.export_node(child)` on the input / output nodes.

Some root Optypes like `Case` and `ExitBlock` should be detected early and rejected here.

Reply:
The core root module node is exported not as a node but as a region. I was not aware that there can be `Hugr`s whose roots are not `Module`s...

Reply:
I propose we postpone this: https://github.com/CQCL/hugr/issues/1554